blockchain/blockchain.go

```diff
@@ -0,0 +1,198 @@
+	Hash   *felt.Felt
+}
+
+func (k *BlockDbKey) MarshalBinary() ([]byte, error) {
+	var numB [8]byte
+	binary.BigEndian.PutUint64(numB[:], k.Number)
+	return db.Blocks.Key(numB[:], k.Hash.Marshal()), nil
+}
+
+func (k *BlockDbKey) UnmarshalBinary(data []byte) error {
+	if len(data) != 41 {
+		return errors.New("key should be 41 bytes long")
+	}
+
+	if data[0] != byte(db.Blocks) {
+		return errors.New("wrong prefix")
+	}
+
+	k.Number = binary.BigEndian.Uint64(data[1:9])
+	k.Hash = new(felt.Felt).SetBytes(data[9:41])
+	return nil
+}
+
+// Blockchain is responsible for keeping track of all things related to the StarkNet blockchain
+type Blockchain struct {
+	l *sync.RWMutex
+
+	head     *core.Block
+	network  utils.Network
+	database db.DB
+}
+
+func NewBlockchain(database db.DB, network utils.Network) (*Blockchain, error) {
+	chain := &Blockchain{
+		l:        &sync.RWMutex{},
+		database: database,
+		network:  network,
+	}
+	txn := chain.database.NewTransaction(false)
+	headBlockBin, err := txn.Get(db.HeadBlock.Key())
+	if errors.Is(err, db.ErrKeyNotFound) {
+		return chain, nil
+	}
+
+	headBlock := new(core.Block)
+	if err = gob.NewDecoder(bytes.NewReader(headBlockBin)).Decode(headBlock); err != nil {
+		return nil, err
+	}
+	chain.head = headBlock
+	return chain, nil
+}
+
+// Height returns the latest block height. If blockchain is empty nil is returned.
+func (b *Blockchain) Height() *uint64 {
+	b.l.RLock()
+	defer b.l.RUnlock()
+	if b.head != nil {
+		h := b.head.Number
+		return &h
+	}
+	return nil
+}
+
+// Store takes a block and state update and performs sanity checks before putting in the database.
+func (b *Blockchain) Store(block *core.Block, stateUpdate *core.StateUpdate) error {
+	if err := b.VerifyBlock(block, stateUpdate); err != nil {
+		return err
+	}
+	if err := state.NewState(b.database).Update(stateUpdate); err != nil {
+		return err
+	}
+
+	key := &BlockDbKey{block.Number, block.Hash}
+	bKey, err := key.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	buffer := new(bytes.Buffer)
+	if err = gob.NewEncoder(buffer).Encode(block); err != nil {
+		return err
+	}
+	blockBinary := append([]byte{}, buffer.Bytes()...)
+
+	if err = b.database.Update(func(txn db.Transaction) error {
+		if err = txn.Set(db.HeadBlock.Key(), blockBinary); err != nil {
+			return err
+		}
+		return txn.Set(bKey, blockBinary)
+	}); err != nil {
+		return err
+	}
+
+	b.l.Lock()
+	b.head = block
+	b.l.Unlock()
+	return nil
+}
+
+func (b *Blockchain) VerifyBlock(block *core.Block, stateUpdate *core.StateUpdate) error {
+	/*
+		Todo: Transaction and TransactionReceipts
+			- When Block is changed to include a list of Transaction and TransactionReceipts
+			- Further checks would need to be added to ensure Transaction Hash has been computed
+				properly.
+			- Sanity check would need to include checks which ensure there is same number of
+				Transactions and TransactionReceipts.
+	*/
+	var head *core.Block
+	b.l.RLock()
+	head = b.head
+	b.l.RUnlock()
```

Comment:
This is open to races. Our local reference to `HEAD` stays the same, but `b.head` could change anytime while we are verifying a block. If it does, we just verified a block for the previous `HEAD`, which then we will try to append on top of the new `HEAD`.

Reply:
Yeah, you are right. I was supposed to create a deep copy but I forgot. I will use the DB to synchronise instead of having mutexes

Reply:
Done